azurerm/resource_arm_operational_insight_workspace.go

```diff
@@ -0,0 +1,207 @@
+			"sku": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(operationalinsights.Free),
+					string(operationalinsights.PerNode),
+					string(operationalinsights.Premium),
+					string(operationalinsights.Standalone),
+					string(operationalinsights.Standard),
+					string(operationalinsights.Unlimited),
+				}, true),
+				DiffSuppressFunc: ignoreCaseDiffSuppressFunc,
+			},
+			"retention_in_days": {
+				Type:         schema.TypeInt,
+				Optional:     true,
+				Computed:     true,
+				ValidateFunc: validateAzureRmOperationalInsightWorkspaceRetentionInDays,
+			},
+			"primary_shared_key": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"secondary_shared_key": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func resourceArmOperationalInsightWorkspaceCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).workspacesClient
+	log.Printf("[INFO] preparing arguments for AzureRM Operational Insight workspace creation.")
+
+	name := d.Get("name").(string)
+	location := d.Get("location").(string)
+	resGroup := d.Get("resource_group_name").(string)
+
+	skuName := d.Get("sku").(string)
+	sku := &operationalinsights.Sku{
+		Name: operationalinsights.SkuNameEnum(skuName),
+	}
+
+	retentionInDays := int32(d.Get("retention_in_days").(int))
+
+	tags := d.Get("tags").(map[string]interface{})
+
+	parameters := operationalinsights.Workspace{
+		Name:     &name,
+		Location: &location,
+		Tags:     expandTags(tags),
+		WorkspaceProperties: &operationalinsights.WorkspaceProperties{
+			Sku:             sku,
+			RetentionInDays: &retentionInDays,
+		},
+	}
+
+	_, error := client.CreateOrUpdate(resGroup, name, parameters, make(chan struct{}))
+	err := <-error
+	if err != nil {
+		return err
+	}
+
+	read, err := client.Get(resGroup, name)
+	if err != nil {
+		return err
+	}
+
+	if read.ID == nil {
+		return fmt.Errorf("Cannot read Operational Inight Workspace '%s' (resource group %s) ID", name, resGroup)
+	}
+
+	d.SetId(*read.ID)
+
+	return resourceArmOperationalInsightWorkspaceRead(d, meta)
+
+}
+
+func resourceArmOperationalInsightWorkspaceRead(d *schema.ResourceData, meta interface{}) error {
+
+	client := meta.(*ArmClient).workspacesClient
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resGroup := id.ResourceGroup
+	name := id.Path["workspaces"]
+
+	resp, err := client.Get(resGroup, name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error making Read request on AzureRM Operational Insight workspaces '%s': %+v", name, err)
+	}
+
+	d.Set("name", resp.Name)
+	d.Set("location", resp.Location)
+	d.Set("resource_group_name", resGroup)
+	d.Set("workspace_id", resp.CustomerID)
+	d.Set("portal_url", resp.PortalURL)
+	d.Set("sku", resp.Sku.Name)
```

Comment:
could we add a check to ensure that the returned SKU object isn't nil here? (this can happen when the Swagger doesn't match the API response, when the SDK is upgraded)

```
if sku := resp.Sku; sku != nil {
  d.Set("sku", sku.Name)
}
```

Reply:
Thanks, I didn't know that!  Good info!

Reply:
Done